ExoPlayer fails with "Source error" on MKV files with compressed content (ContentCompAlgo)

### This issue respects the following points:

- [x] This issue is **not** already reported on [GitHub](https://github.com/jellyfin/jellyfin-android/issues?q=is%3Aopen+is%3Aissue) _(I've searched it)_.
- [x] I agree to follow Jellyfin's [Code of Conduct](https://jellyfin.org/docs/general/community-standards.html#code-of-conduct).
- [x] This report addresses only a single issue; If you encounter multiple issues, kindly create separate reports for each one.

### Describe the bug

1. Find an MKV file with embedded compressed PGS subtitles
2. Switch to Exoplayer in the "Client Settings"
3. Try to play

Expected: the content plays
Actual: a "Source Error" is shown and no media plays. Enabling subtitles causes the server to transcode and that also mitigates the issue.

Digging in, it looks like this is due to the PGS subitles in the MKV container causing DirectPlay to fail. On Android TV it seems to fall back to DirectStream if this occurs.

### Media info of the file

```shell
Video
Title: 1080p H264 SDR
Codec: H264
AVC: Yes
Profile: High
Level: 40
Resolution: 1920x1080
Aspect ratio: 16:9
Anamorphic: No
Interlaced: No
Framerate: 23.976025
Bitrate: 11252 kbps
Bit depth: 8 bit
Video range: SDR
Video range type: SDR
Pixel format: yuv420p
Ref frames: 1
NAL: 4


Audio
Title: English - Dolby Digital - 5.1 - Default
Language: eng
Codec: AC3
AVC: No
Layout: 5.1
Channels: 6 ch
Bitrate: 448 kbps
Sample rate: 48000 Hz
Default: Yes
Forced: No
External: No


Subtitle
Title: English - Default - DVDSUB
Language: eng
Codec: DVDSUB
AVC: No
Resolution: 1920x1080
Default: Yes
Forced: No
External: No
```

### Which video player implementations does this bug apply to?

- [ ] Web player (default)
- [x] Integrated player (ExoPlayer)
- [ ] External player (VLC, mpv, MX Player)

### Client logs

```shell
03-15 23:40:05.130  8042  8042 E EventLogger: internalError [eventTime=0.23, mediaPos=5106.57, window=0, period=0, loadError
03-15 23:40:05.130  8042  8042 E EventLogger:   androidx.media3.common.ParserException: ContentCompAlgo 0 not supported {contentIsMalformed=true, dataType=1}
03-15 23:40:05.135  8042  8192 E ExoPlayerImplInternal: Playback error
03-15 23:40:05.135  8042  8192 E ExoPlayerImplInternal:   androidx.media3.exoplayer.ExoPlaybackException: Source error
03-15 23:40:05.135  8042  8192 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.ExoPlayerImplInternal.handleIoException(ExoPlayerImplInternal.java:927)
03-15 23:40:05.135  8042  8192 E ExoPlayerImplInternal:   Caused by: androidx.media3.common.ParserException: ContentCompAlgo 0 not supported {contentIsMalformed=true, dataType=1}
03-15 23:40:05.137  8042  8042 E EventLogger: playerFailed [eventTime=0.24, mediaPos=5106.57, window=0, period=0, errorCode=ERROR_CODE_PARSING_CONTAINER_MALFORMED
03-15 23:40:05.137  8042  8042 E EventLogger:   androidx.media3.exoplayer.ExoPlaybackException: Source error
03-15 23:40:05.137  8042  8042 E EventLogger:       at androidx.media3.exoplayer.ExoPlayerImplInternal.handleIoException(ExoPlayerImplInternal.java:927)
03-15 23:40:05.137  8042  8042 E EventLogger:   Caused by: androidx.media3.common.ParserException: ContentCompAlgo 0 not supported {contentIsMalformed=true, dataType=1}
```

### FFmpeg logs

```shell
I don't see any ffmpeg logs, likely because this is invoking DirectPlay.
```

### Application version

2.6.3

### Where did you install the app from?

Google Play Store

### Device information

Google Pixel 9 Pro

### Android version

Android 16

### Jellyfin server version

10.11.6

### Other sources

_No response_

--- a/app/src/main/java/org/jellyfin/mobile/player/PlayerViewModel.kt
+++ b/app/src/main/java/org/jellyfin/mobile/player/PlayerViewModel.kt
@@ -141,6 +141,7 @@ class PlayerViewModel(application: Application) : AndroidViewModel(application),
     private var progressUpdateJob: Job? = null
     private var chapterMarkingUpdateJob: Job? = null
     private var skipMediaSegmentUpdateJob: Job? = null
+    private var fallbackRetryJob: Job? = null
 
     /**
      * Returns the current ExoPlayer instance or null
@@ -776,8 +777,21 @@ class PlayerViewModel(application: Application) : AndroidViewModel(application),
             setupPlayer()
             queueManager.tryRestartPlayback()
         } else {
-            _error.postValue(error.localizedMessage.orEmpty())
+            Timber.w(error, "Playback error, attempting fallback")
+            val startPosition = (playerOrNull?.currentPosition ?: 0L).milliseconds
+            fallbackRetryJob?.cancel()
+            fallbackRetryJob = viewModelScope.launch {
+                val retried = queueManager.restartPlaybackWithFallback(startPosition)
+                if (!retried) {
+                    _error.postValue(error.localizedMessage.orEmpty())
+                }
+            }
         }
+    }
+
+    fun cancelFallbackRetry() {
+        fallbackRetryJob?.cancel()
+        fallbackRetryJob = null
     }
 
     override fun onCleared() {

--- a/app/src/main/java/org/jellyfin/mobile/player/queue/QueueManager.kt
+++ b/app/src/main/java/org/jellyfin/mobile/player/queue/QueueManager.kt
@@ -37,6 +37,7 @@ import org.jellyfin.sdk.model.serializer.toUUIDOrNull
 import org.koin.core.component.KoinComponent
 import org.koin.core.component.get
 import org.koin.core.component.inject
+import timber.log.Timber
 import java.util.UUID
 import kotlin.time.Duration
 
@@ -53,6 +54,9 @@ class QueueManager(
     private var currentQueue: List<UUID> = emptyList()
     private var currentQueueIndex: Int = 0
 
+    private var playbackRetries = 0
+    private var lastPlaybackError = 0L
+
     private val _currentMediaSource: MutableLiveData<JellyfinMediaSource> = MutableLiveData()
     val currentMediaSource: LiveData<JellyfinMediaSource>
         get() = _currentMediaSource
@@ -68,6 +72,7 @@ class QueueManager(
     suspend fun initializePlaybackQueue(playOptions: PlayOptions): PlayerException? {
         currentQueue = playOptions.ids
         currentQueueIndex = playOptions.startIndex
+        resetPlaybackFallback()
 
         val itemId = when {
             currentQueue.isNotEmpty() -> currentQueue[currentQueueIndex]
@@ -145,6 +150,8 @@ class QueueManager(
         audioStreamIndex: Int? = null,
         subtitleStreamIndex: Int? = null,
         playWhenReady: Boolean = true,
+        enableDirectPlay: Boolean? = null,
+        enableDirectStream: Boolean? = null,
     ): PlayerException? {
         mediaSourceResolver.resolveMediaSource(
             itemId = itemId,
@@ -154,6 +161,8 @@ class QueueManager(
             startTime = startTime,
             audioStreamIndex = audioStreamIndex,
             subtitleStreamIndex = subtitleStreamIndex,
+            enableDirectPlay = enableDirectPlay,
+            enableDirectStream = enableDirectStream,
         ).onSuccess { jellyfinMediaSource ->
             // Ensure transcoding of the current element is stopped
             getCurrentMediaSourceOrNull()?.let { oldMediaSource ->
@@ -184,6 +193,55 @@ class QueueManager(
                 viewModel.load(this, it, playWhenReady = true)
             }
         }
+    }
+
+    private fun resetPlaybackFallback() {
+        playbackRetries = 0
+        lastPlaybackError = 0L
+        viewModel.cancelFallbackRetry()
+    }
+
+    /**
+     * Retry playback with progressively degraded settings to work around source errors.
+     *
+     * Retry 1: no constraints — lets the server decide freely.
+     * Retry 2: disable direct play, allowing the server to fall back to direct stream.
+     * Retry 3: disable direct stream too, forcing the server to transcode.
+     *
+     * @param startPosition The position at which to resume playback after the retry.
+     * @return true if a retry was initiated, false if retries are exhausted or not applicable.
+     */
+    suspend fun restartPlaybackWithFallback(startPosition: Duration): Boolean {
+        val currentMediaSource = getCurrentMediaSourceOrNull() as? RemoteJellyfinMediaSource ?: return false
+
+        val now = System.currentTimeMillis()
+        if (playbackRetries > 0 && now - lastPlaybackError > PLAYBACK_RETRY_RESET_MS) {
+            Timber.i("Playback stabilized, resetting retry count from %d", playbackRetries)
+            playbackRetries = 0
+        }
+
+        playbackRetries++
+        lastPlaybackError = now
+
+        if (playbackRetries > MAX_PLAYBACK_RETRIES) return false
+
+        Timber.i("Retrying playback (attempt %d of %d)", playbackRetries, MAX_PLAYBACK_RETRIES)
+
+        // If already transcoding, only retry once (for transient errors); flags are null on
+        // retry 1 anyway, so no special-casing needed in the startRemotePlayback call.
+        if (currentMediaSource.playMethod == PlayMethod.TRANSCODE && playbackRetries > 1) return false
+
+        return startRemotePlayback(
+            itemId = currentMediaSource.itemId,
+            mediaSourceId = currentMediaSource.id,
+            maxStreamingBitrate = currentMediaSource.maxStreamingBitrate,
+            startTime = startPosition,
+            audioStreamIndex = currentMediaSource.selectedAudioStreamIndex,
+            subtitleStreamIndex = currentMediaSource.selectedSubtitleStreamIndex,
+            playWhenReady = true,
+            enableDirectPlay = if (playbackRetries > 1) false else null,
+            enableDirectStream = if (playbackRetries > 2) false else null,
+        ) == null
     }
 
     /**
@@ -217,6 +275,8 @@ class QueueManager(
 
         val currentMediaSource = getCurrentMediaSourceOrNull() as? RemoteJellyfinMediaSource ?: return false
 
+        resetPlaybackFallback()
+
         startRemotePlayback(
             itemId = currentQueue[--currentQueueIndex],
             mediaSourceId = null,
@@ -227,6 +287,8 @@ class QueueManager(
 
     suspend fun next(): Boolean {
         if (!hasNext()) return false
+
+        resetPlaybackFallback()
 
         when (val currentMediaSource = getCurrentMediaSourceOrNull()) {
             is LocalJellyfinMediaSource -> startDownloadPlayback(
@@ -371,6 +433,7 @@ class QueueManager(
     suspend fun selectAudioStreamAndRestartPlayback(stream: MediaStream): Boolean {
         require(stream.type == MediaStreamType.AUDIO)
         val currentPlayState = viewModel.getStateAndPause() ?: return false
+        resetPlaybackFallback()
 
         when (val currentMediaSource = getCurrentMediaSourceOrNull()) {
             is LocalJellyfinMediaSource -> startDownloadPlayback(
@@ -404,6 +467,7 @@ class QueueManager(
     suspend fun selectSubtitleStreamAndRestartPlayback(stream: MediaStream?): Boolean {
         require(stream == null || stream.type == MediaStreamType.SUBTITLE)
         val currentPlayState = viewModel.getStateAndPause() ?: return false
+        resetPlaybackFallback()
 
         when (val mediaSource = getCurrentMediaSourceOrNull()) {
             is LocalJellyfinMediaSource -> startDownloadPlayback(
@@ -425,5 +489,10 @@ class QueueManager(
             null -> return false
         }
         return true
+    }
+
+    companion object {
+        private const val MAX_PLAYBACK_RETRIES = 3
+        private const val PLAYBACK_RETRY_RESET_MS = 30_000L
     }
 }

--- a/app/src/main/java/org/jellyfin/mobile/player/source/MediaSourceResolver.kt
+++ b/app/src/main/java/org/jellyfin/mobile/player/source/MediaSourceResolver.kt
@@ -32,6 +32,8 @@ class MediaSourceResolver(private val apiClient: ApiClient) {
         audioStreamIndex: Int? = null,
         subtitleStreamIndex: Int? = null,
         autoOpenLiveStream: Boolean = true,
+        enableDirectPlay: Boolean? = null,
+        enableDirectStream: Boolean? = null,
     ): Result<RemoteJellyfinMediaSource> {
         // Load media source info
         val playSessionId: String
@@ -50,6 +52,8 @@ class MediaSourceResolver(private val apiClient: ApiClient) {
                         audioStreamIndex = audioStreamIndex,
                         subtitleStreamIndex = subtitleStreamIndex,
                         autoOpenLiveStream = autoOpenLiveStream,
+                        enableDirectPlay = enableDirectPlay,
+                        enableDirectStream = enableDirectStream,
                     ),
                 ).content
             }

